torch/csrc/autograd/python_function.cpp

```diff
@@ -1017,11 +1017,11 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
   if (!self) {
     return std::shared_ptr<PyFunction>();
   }
-  Py_INCREF((PyObject*)self);
 
   auto ptr = self->cdata_ptr.lock();
   if (ptr) return ptr;
 
+  Py_INCREF((PyObject*)self);
```

Comment:
Because the `shared_ptr` uses the `Decref` as the deleter. Once all shared_ptrs are freed, it's `operator()` will be called and will revert this incref.